### ISSUE TITLE
Plotting bug when using push_notebook() from CustomJS callback

### ISSUE THREAD

**Issue by jdoepfert** [bokeh/bokeh#5452]
I am trying to update a bokeh plot with a bokeh slider widget from within a jupyter notebook using the following code:

```
from bokeh.plotting import figure
from bokeh.io import output_notebook, push_notebook, show
from bokeh.models import CustomJS, Slider

output_notebook()

power = 0.5
x = [1,2,3]
y = [i**power for i in x]

fig = figure()
plt = fig.circle(x, y)

def update_plot(power):
    x = plt.data_source.data['x']
    plt.data_source.data['y'] = [i**power for i in x]
    push_notebook(handle=bokeh_handle)  


bokeh_handle = show(fig, notebook_handle=True)

##### new notebook cell #####

callback = CustomJS(code="""
if (IPython.notebook.kernel !== undefined) {
    var kernel = IPython.notebook.kernel;
    cmd = "update_plot(" + cb_obj.value + ")";
    kernel.execute(cmd, {}, {});
}
""")

slider = Slider(start=0.1, end=1, value=1, step=.05, title="power", callback=callback)
show(slider)
```
The idea is that the JS callback for the slider calls the python function update_plot(), which changes the data of the bokeh plot and then triggers a push_notebook().

However, when I move the slider, the plot is not updated, but  a mini version of the plot appears in the upper left corner (see red arrow [here](https://i.stack.imgur.com/AiZlm.png)).

Executing `print(plt.data_source.data['y'])` showed me that the callback and `update_plot()` were actually called upon slider movement. 

@bryevdv suggests the following reason [here](http://stackoverflow.com/questions/40533394/updating-bokeh-plot-with-a-bokeh-widget-in-jupyter-notebook):

> it seems like there might be a problem with update events (i.e. push_notebook) coming mid-flight during another ongoing event processing (the slider callback)

* Software version information output by `bokeh info` command:
Python version      :  3.5.2 |Anaconda 4.1.1 (64-bit)| (default, Jul  2 2016, 17:53:06) 
IPython version     :  5.1.0
Bokeh version       :  0.12.3
BokehJS static path :  /home/jefe/anaconda3/lib/python3.5/site-packages/bokeh/server/stati

**Comment by bryevdv** [comment]
Marking `discussion` pending further investigation. It may that this is simply an inherent limitation we have to document.

**Comment by kdheepak** [comment]
> it seems like there might be a problem with update events (i.e. push_notebook) coming mid-flight during another ongoing event processing (the slider callback)

Hi @bryevdv, are you sure that this is the case? I tested the following and even after "ending callback" was logged to the console and calling the `update_plot` function after 5 seconds, I was still experiencing the same issue.

``` python
callback = CustomJS(code="""
function delayed(){
    if (IPython.notebook.kernel !== undefined) {
        var kernel = IPython.notebook.kernel;
        cmd = "update_plot(" + cb_obj.value + ")";
        kernel.execute(cmd, {}, {});
    }
}
console.log('starting callback')
setTimeout(function() { delayed(); }, 5000);
console.log('ending callback')
""")
```
I would really like this feature too, since this will allow working entirely with Bokeh widgets in the notebook. This will make prototyping Bokeh server applications via the notebook straightforward, hence making the transition from notebook exploration to application seamless.

**Comment by jdoepfert** [comment]
Plot updating seems to work correctly when displaying the plot using
```
from bokeh.layouts import row
bokeh_handle = show(row(fig, slider), notebook_handle=True)
```
instead of 
`bokeh_handle = show(fig, notebook_handle=True)` and `show(slider)`.
Maybe this helps @kdheepak

**Comment by bryevdv** [comment]
This appears to be fixed with #6945 testing with the code below
```
from bokeh.plotting import figure
from bokeh.io import output_notebook, push_notebook, show
from bokeh.models import CustomJS, Slider

output_notebook()

power = 0.5
x = [1,2,3,4,5,6]
y = [i**power for i in x]

fig = figure()
plt = fig.circle(x, y)

def update_plot(power):
    x = plt.data_source.data['x']
    plt.data_source.data['y'] = [i**power for i in x]
    push_notebook(handle=bokeh_handle)  


##### new notebook cell #####

callback = CustomJS(code="""
if (IPython.notebook.kernel !== undefined) {
    var kernel = IPython.notebook.kernel;
    cmd = "update_plot(" + cb_obj.value + ")";
    kernel.execute(cmd, {}, {});
}
""")

slider = Slider(start=0.1, end=1, value=1, step=.05, title="power", callback=callback)
from bokeh.layouts import column
bokeh_handle = show(column(fig, slider), notebook_handle=True)
```
